File: leetcode-js/017-letter-combinations-of-a-phone-number.js
var table = [
    ' ',
    '',     'abc', 'def',
    'ghi',  'jkl', 'mno',
    'pqrs', 'tuv', 'wxyz'
];

/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    if (!digits.length) {
        return [];
    }

    const result = [];
    findCombination('', digits, 0, result);
    return result;
};

function findCombination(str, digits, i, result) {
    'use strict';

    if (i >= digits.length) {
        result.push(str);
        return;
    }

    const letters = table[digits[i]];

    if (!letters.length) {
        findCombination(str, digits, i + 1, result);
        return;
    }

    for (let j = 0; j < letters.length; j += 1) {
        findCombination(str + letters[j], digits, i + 1, result);
    }
}

File: leetcode-java/017-letter-combinations-of-a-phone-number.java
public class Solution {
    public List<String> letterCombinations(String digits) {
        LinkedList<String> res = new LinkedList<>();

        if (digits.length() == 0) {
            return res;
        }

        String[] mapping = new String[] {"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

        res.add("");

        for (int i = 0; i < digits.length(); i++) {
            int n = Character.getNumericValue(digits.charAt(i));
            char[] chars = mapping[n].toCharArray();

            while (res.peek().length() == i) {
                String str = res.poll();

                for (char c : chars) {
                    res.offer(str + c);
                }
            }
        }

        return res;
    }
}
